vscode.fhs: add buildFHSUserEnv wrapped version

jonringer opened the pull request:
###### Motivation for this change
Extensions commonly use pre-compiled binaries which aren't
aware of the nix store. Using buildFHSUserEnv will allow
for a more "natural" usage of vscode.

EDIT:
To not break users due to FHSUser regressions, I'm instead adding this as a passthru to the original vscode/vscodium flavors.

closes: #73810
closes: #91179

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

buckley310 commented:
This opens and mostly works for me, although unfortunately the UserEnv also applies to vscode's integrated terminal, which would basically break that feature for me. Not sure if that can be fixed.

mebubo commented:
I'd like to keep using the unwrapped version. Would it be possible to make this overridable?

jonringer commented:
> I'd like to keep using the unwrapped version. Would it be possible to make this overridable?

it's on the passthru `vscode.unwrapped`

jonringer commented:
I'm in no hurry to merge this, and it could probably be improved to support the other use cases. Meant for this to be more of a discussion than anything else

buckley310 commented:
Using the unwrapped version seems good enough for me since I already got my plugins working natively.
It does seem desirable to have plugins working out of the box, though. That did cause me some trouble early on.

I wonder if it would be a good idea for discoverability if the unwrapped versions of vscode/vscodium were added to `all-packages`, so that they would show up on https://search.nixos.org/packages

jonringer commented:
the other option would be to expose `vscode-fhs`, but I'm not a big fan of that.

@buckley310 can you give some steps to repro the terminal breaking?

buckley310 commented:
I have gcc in my systemPackages, but when I try and compile a C hello-world from the wrapped vscode terminal gcc returns
```
ld: cannot find crt1.o: No such file or directory
collect2: error: ld returned 1 exit status
```

The bash prompt itself is just the bash default `bash-4.4$`, some hotkeys (ctrl+arrows) and my aliases were not working as well since a lot of the global bashrc stuff doesn't apply to UserEnvs. Some or all of that might be fixable via ~/.bashrc, but I haven't tried.

jonringer commented:
I'll try to see what I can do, it would be nice for vscode to "just work"

jonringer commented:
@buckley310 do you mind  trying again. I was able to get my terminal to running fine

jonringer commented:
@Atemu  @illegalprime do you mind reviewing these changes?

buckley310 commented:
Bash interaction seems all good now.
Gcc issue is still present.
I actually noticed that SSL certificates seem to be problematic now, but maybe that one is just a missing bind in /etc/.

jonringer commented:
what gcc issues are you having?

buckley310 commented:
```
echo 'int main(){return 0;}' >a.c ; gcc a.c
```
fails with
```
/nix/store/hiwz81i1g3fn3p0acjs042a4h5fri6dh-binutils-2.31.1/bin/ld: cannot find crt1.o: No such file or directory
/nix/store/hiwz81i1g3fn3p0acjs042a4h5fri6dh-binutils-2.31.1/bin/ld: cannot find crti.o: No such file or directory
collect2: error: ld returned 1 exit status
```

jonringer commented:
@buckley310 I figured it out, the unwrapped gcc gets put into the environment

jonringer commented:
the `libstdc++.so` seems to be present even if I remove any mention of stdenv.

@buckley310 please try again

jonringer commented:
doing `stdenv.cc` in targetPkgs does make gcc-wrapped available, but I (for example) don't normally have gcc in my path

jonringer commented:
since I have buildFHSUserEnv sourcing your bash terminal. It should be inheriting from your normal shell

jonringer commented:
pinging vscode extension contributors @eadwu @raboof @mcwitt  @austinbutler @oxalica @veprbl 

do you all mind testing this?

If you normally track unstable, this should be something like:
```
home-manager -I nixpkgs=https://github.com/jonringer/nixpkgs/archive/vscode-buildfhsuserenv.tar.gz switch
```

veprbl commented:
@jonringer Just in case, I'm not a vscode user. I was only reviewing vscode PR's because it used to be that there was no one else to look at those.

jonringer commented:
oh, then thank you for your contributions. Sorry for bothering you :)

jonringer commented:
wow, burned a lot of time with https://github.com/NixOS/nixpkgs/pull/101265

jonringer commented:
the MITSHIM is only needed because we imply `--unshare-ipc` from `--unshare-all` in buildFHSUserEnvBubblewrap.
Maybe instead, we should only unshare the other items

austinbutler commented:
@jonringer I use `home-manager` as a NixOS module so don't have `home-manager` directly available. I added overrides for `vscode` and `vscode-extensions` (not sure if both are necessary or if one will get in the way) in the way I normally override packages, though.

```nix
{
  nixpkgs.overlays = [
    (self: super: {
      vscode = (import (builtins.fetchTarball {
        url =
          "https://github.com/jonringer/nixpkgs/archive/34dc29b45c5f560c32b65937c265016f526be9fc.tar.gz";
        sha256 = "0p90nlia627hqn14rrs4bpbpn4bb1mjicddhc7av9lpkcy5c3l4h";
      }) { config = { allowUnfree = true; }; }).vscode;

      vscode-extensions = (import (builtins.fetchTarball {
        url =
          "https://github.com/jonringer/nixpkgs/archive/34dc29b45c5f560c32b65937c265016f526be9fc.tar.gz";
        sha256 = "0p90nlia627hqn14rrs4bpbpn4bb1mjicddhc7av9lpkcy5c3l4h";
      }) { config = { allowUnfree = true; }; }).vscode-extensions;
    })
  ];
}
```

When I open VS Code after switching I get a popup about Git not being found and the terminal is using a basic ZSH instead of my normal powerline goodies, aliases, etc. FYI I do have a number of VSC Extension installed via home-manager.

jonringer commented:
> @jonringer I use `home-manager` as a NixOS module so don't have `home-manager` directly available. I added overrides for `vscode` and `vscode-extensions` (not sure if both are necessary or if one will get in the way) in the way I normally override packages, though.
> 
> ```nix
> {
>   nixpkgs.overlays = [
>     (self: super: {
>       vscode = (import (builtins.fetchTarball {
>         url =
>           "https://github.com/jonringer/nixpkgs/archive/34dc29b45c5f560c32b65937c265016f526be9fc.tar.gz";
>         sha256 = "0p90nlia627hqn14rrs4bpbpn4bb1mjicddhc7av9lpkcy5c3l4h";
>       }) { config = { allowUnfree = true; }; }).vscode;
> 
>       vscode-extensions = (import (builtins.fetchTarball {
>         url =
>           "https://github.com/jonringer/nixpkgs/archive/34dc29b45c5f560c32b65937c265016f526be9fc.tar.gz";
>         sha256 = "0p90nlia627hqn14rrs4bpbpn4bb1mjicddhc7av9lpkcy5c3l4h";
>       }) { config = { allowUnfree = true; }; }).vscode-extensions;
>     })
>   ];
> }
> ```
> 
> When I open VS Code after switching I get a popup about Git not being found and the terminal is using a basic ZSH instead of my normal powerline goodies, aliases, etc. FYI I do have a number of VSC Extension installed via home-manager.



> ```nix
> ```nix
> 
> ```
> 
> 
> When I open VS Code after switching I get a popup about Git not being found and the terminal is using a basic ZSH instead of my normal powerline goodies, aliases, etc. FYI I do have a number of VSC Extension installed via home-manager.
> ```

do you mind displaying your zsh related `/etc/` files?

to get bash working, i had to ro-bind `/etc/bashrc` into the chroot'd environment.

austinbutler commented:
These are the zsh-related files in `/etc/`.

`/etc/zshenv`

```sh
# /etc/zshenv: DO NOT EDIT -- this file has been generated automatically.
# This file is read for all shells.

# Only execute this file once per shell.
if [ -n "$__ETC_ZSHENV_SOURCED" ]; then return; fi
__ETC_ZSHENV_SOURCED=1

if [ -z "$__NIXOS_SET_ENVIRONMENT_DONE" ]; then
    . /nix/store/w4nvlk6vh8mg6zr7wjkk5zp7ya2ms7jq-set-environment
fi

HELPDIR="/nix/store/a9f0q1cg0x2xj1pvvp52l86n8a531y3b-zsh-5.8/share/zsh/$ZSH_VERSION/help"

# Tell zsh how to find installed completions.
for p in ${(z)NIX_PROFILES}; do
    fpath+=($p/share/zsh/site-functions $p/share/zsh/$ZSH_VERSION/functions $p/share/zsh/vendor-completions)
done

# Setup custom shell init stuff.




# Read system-wide modifications.
if test -f /etc/zshenv.local; then
    . /etc/zshenv.local
fi
```

`/etc/zshrc`

```sh
# /etc/zshrc: DO NOT EDIT -- this file has been generated automatically.
# This file is read for interactive shells.
#
# Note that generated /etc/zprofile and /etc/zshrc files do a lot of
# non-standard setup to make zsh usable with no configuration by default.
#
# Which means that unless you explicitly meticulously override everything
# generated, interactions between your ~/.zshrc and these files are likely
# to be rather surprising.
#
# Note however, that you can disable loading of the generated /etc/zprofile
# and /etc/zshrc (you can't disable loading of /etc/zshenv, but it is
# designed to not set anything surprising) by setting `no_global_rcs` option
# in ~/.zshenv:
#
#   echo setopt no_global_rcs >> ~/.zshenv
#
# See "STARTUP/SHUTDOWN FILES" section of zsh(1) for more info.


# Only execute this file once per shell.
if [ -n "$__ETC_ZSHRC_SOURCED" -o -n "$NOSYSZSHRC" ]; then return; fi
__ETC_ZSHRC_SOURCED=1

# Set zsh options.
setopt HIST_IGNORE_DUPS SHARE_HISTORY HIST_FCNTL_LOCK


# Setup command line history.
# Don't export these, otherwise other shells (bash) will try to use same HISTFILE.
SAVEHIST=2000
HISTSIZE=2000
HISTFILE=$HOME/.zsh_history

# Configure sane keyboard defaults.
. /etc/zinputrc

# Enable autocompletion.
autoload -U compinit && compinit




# Setup custom interactive shell init stuff.


# oh-my-zsh configuration generated by NixOS
export ZSH=/nix/store/h1yc1l6n98dc3hsrwgsmhzy3gga7gqxq-oh-my-zsh-2020-10-18/share/oh-my-zsh
```

And my related config from `configuration.nix`:
```nix
{
  users.defaultUserShell = pkgs.zsh;
  programs.zsh = {
    enable = true;
    ohMyZsh.enable = true;
    promptInit =
      "source ${pkgs.zsh-powerlevel10k}/share/zsh-powerlevel10k/powerlevel10k.zsh-theme";
    shellAliases = {
      nixgc = "sudo nix-collect-garbage -d";
      nixup = "cd ~/Documents/nixos && bash upgrade.sh && cd -";
    };
    vteIntegration = true;
  };
}
```

jonringer commented:
@austinbutler do you mind trying again? :)

buckley310 commented:
`curl https://x.org` does not work on 34dc29b. certificate errors. That's the only problem left that I can find.
It is a little quirky that vscode is in it's own PID namespace now, but I don't think that would really bother me. Maybe others if they spend a lot more time in the integrated terminal?

jonringer commented:
@buckley310 that probably because we do `--unshare-all` to the chroot, I will probably make this more "configurable" to allow a "barely" isolated vscode. But I have to go cook dinner

austinbutler commented:
> @austinbutler do you mind trying again? :)

I still get the "Git not found" error when I open VSC. The terminal now looks a bit more like my normal terminal, but still it is not the same (no git status). And the terminal has a new error: `/etc/zshrc:.:36: no such file or directory: /etc/zinputrc`.

austinbutler commented:
Maybe you need to also add `/etc/zinputrc` and `/etc/zprofile`.

jonringer commented:
never realised how distributed zsh files were

austinbutler commented:
Latest commit has resolved the `zinputrc` error, but still Git is not found. I do have my ZSH aliases though.

jonringer commented:
> Latest commit has resolved the `zinputrc` error, but still Git is not found. I do have my ZSH aliases though.

how do you have git installed?

I have it configured through home-manager, and it's able to be found

jonringer commented:
I might break this out into a separate top-level attr. Something like `vscode-fhs` or added it to the passthru as `vscode.fhs`

Actually, just having a "fhs" on the original vscode/vscodium might be best.

thoughts? @Atemu

jonringer commented:
@timokau how do you feel about instead of making `vscode` wrapped, just provide a `vscode.fhs` "variant"

cc @FRidh , probably know of a better way

timokau commented:
Did you ping me on purpose? I have never contributed to or used vscode.

jonringer commented:
I did, I was asking more about how this should be exposed in nixpkgs.

Currently the PR is replacing the existing vscode/vscodium, but there's still some regressions due to how our currently buildFHSUserEnvBubblewrap is called. And I was wondering if we should just provide a "new" attr, so that people using the previous workflow aren't "suddenly" broken.

FRidh commented:
I prefer to avoid using `buildFHSUserEnv`, because its more of a last-resort method, but I understand why some would like it. For that reason I would prefer to be able to choose.

jonringer commented:
> I prefer to avoid using `buildFHSUserEnv`, because its more of a last-resort method, but I understand why some would like it. For that reason I would prefer to be able to choose.

Being able to use vscode extensions with no additional effort is a pretty big win for some (myself included, where some projects just say to inherit the vscode settings to avoid linting and styling CI pain)

I'll refactor the package so there's a `fhs` attr on the `passthru`. And  people can just do `programs.vscode.package = pkgs.vscode.fhs;` in home-manager to bring it in

timokau commented:
> I did, I was asking more about how this should be exposed in nixpkgs.

Ah, I see. My uneducated opinion is that it would be nice to have the "just works" (FHS) option as the default, with the option to chose the plain version. If its possible to detect broken behavior in the FHS version, you could print a warning and suggest that uses switch to the plain version. If that is not possible its maybe better to keep the plain version as the default now.

But I'm really not very qualified to comment on this :)

jonringer commented:
> But I'm really not very qualified to comment on this :)

Well, it's still nice to get feedback, in case you're going down the wrong path. Thanks for your input.

The FHS "regressions" are more around creating kernel namespaces, which makes it feel like you're developing on "someone else's machine" rather than your own.

I think I will use the `vscode.fhs` path for now, and then when we are "stable enough" maybe switch it.

Atemu commented on pkgs/applications/editors/vscode/generic.nix:
```diff
@@ -1,7 +1,7 @@
 , unzip, libsecret, libXScrnSaver, wrapGAppsHook
 , gtk2, atomEnv, at-spi2-atk, autoPatchelfHook
-, systemd, fontconfig, libdbusmenu
+, systemd, fontconfig, libdbusmenu, buildFHSUserEnvBubblewrap
```
Nit: `buildFHSUserEnvBubblewrap` is only meant as a temporary measure until it replaces `buildFHSUserEnv` completely, so vscode should probably not explicitly depend on it and rather override the `buildFHSUserEnv` with it in its callPackage.

jonringer replied:
I was thinking about this, but I'm starting to like having dependency mapping in two places.

Atemu reviewed (approved):
Bubblewrap changes LGTM. All FHS packages I use still work just like before.

Can't comment on vscode, I use Emacs :p

Atemu commented:
The . in `vscode.fhs` kind of implies that `fhs` is an attribute of `vscode` and the only attributes I'd expect under a derivation are mkDerivation arguments and other implementation details. I think the two variants should be on the same level.  

My proposal would be either: Add a `vscode-fhs` package or switch vscode over to FHS by default and provide -unwrapped.

In both cases there should be an explanation why there are two packages and what makes them different from another in the desciprion so that it show up in the package search. Many users will have no idea what -unwrapped or -fhs stand for (first hand experience).

jonringer commented:
> The . in `vscode.fhs` kind of implies that `fhs` is an attribute of `vscode` and the only attributes I'd expect under a derivation are mkDerivation arguments and other implementation details. I think the two variants should be on the same level.

part of this too, is that vscodium would also received a wrapped version. I would have to do another `vscodium-fhs` to be consistent. I would rename `fhs` to `fhs-wrapped` if that's more explicit as to what's going on

most C compilers have a `.cc` attr which points to the unwrapped version, it's not the most common paradigm, but it does exist.

austinbutler commented:
> > Latest commit has resolved the `zinputrc` error, but still Git is not found. I do have my ZSH aliases though.
> 
> how do you have git installed?
> 
> I have it configured through home-manager, and it's able to be found

In home manager config in `configuration.nix`:
```
programs.git.enable = true;
```

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/new-users-using-nixos-wrong/9973/38

Commit:
buildFHSUserEnvBubblewrap: add dieWithParent option, and /etc/nix

Allows for processes which fork to not be immediately
killed when the parent process dies.

Commit:
vscod{e,ium}: Add fhs passthru

jonringer commented:
decided to revive this. You can do
```
vscode.fhs
# or
vscode.fhsWithPackages (pkgs: with pkgs; [ rustup zlib ]; )
```
to either have a minimal environment or one with additional tools that you only care to have a available for vscode.

Been able to use extensions and the terminal fine

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/500

jonringer commented on pkgs/applications/editors/vscode/generic.nix:
```diff
@@ -94,4 +96,58 @@ in
+      done
+    '';
+
+    runScript = "${unwrapped}/bin/code";
```
good point

jonringer replied:
done. Thanks

jonringer commented:
cc @FRidh @SuperSandro2000 

In the current iteration, the existing behavior is preserved. People are able to use the `fhs` attr to opt-into the fhs environment. I think this is good to merge.

SuperSandro2000 commented:
I can't say anything about this because I don't use vscode in nix or know to much about bubblewrap.

FRidh commented:
I saw you mentioned it as an alternative before but I did not comment that earlier. For discoverability you may want to make it a top-level package instead of a `passthru attribute` (`vscode-fhs`?) and extend or replace the description with a note saying how it is different. Both options are fine in my opinion.

Commit:
vscod{e,ium}-fhs: add top-level aliases, add description

jonringer commented:
> I saw you mentioned it as an alternative before but I did not comment that earlier. For discoverability you may want to make it a top-level package instead of a `passthru attribute` (`vscode-fhs`?) and extend or replace the description with a note saying how it is different. Both options are fine in my opinion.

did both :)

turion commented on pkgs/applications/editors/vscode/generic.nix:
```diff
@@ -97,4 +99,64 @@ in
+
+    meta = meta // {
+      description = ''
+        Wrapped variant of ${pname} which launches in a FHS compatiable envrionment.
```
```suggestion
        Wrapped variant of ${pname} which launches in a FHS compatible environment.
```

jonringer replied:
thanks :)

turion commented on pkgs/applications/editors/vscode/generic.nix:
```diff
@@ -97,4 +99,64 @@ in
+    };
+
+    meta = meta // {
+      description = ''
```
I think it would be better if this description didn't replace the original, but extended it by this paragraph. Also, the descriptions of the unwrapped packages should mention that this FHS wrapped package exist, and hint at the typical kinds of bugs in which one would want to opt for it.

jonringer replied:
> I think it would be better if this description didn't replace the original, but extended it by this paragraph.

> the descriptions of the unwrapped packages should mention that this FHS wrapped package exist

I think this would be awkward in practice, as if I extended the unwrapped version which contains a reference to the fhs version, then the fhs version would mention itself then detail how it's different.

I think this is mostly subjective. If someone was searching using https://search.nixos.org/packages?channel=unstable&from=0&size=50&sort=relevance&query=vscode , then they would find both results. And I think the current description would make the distinction clearer.

turion replied:
> I think this is mostly subjective. If someone was searching using https://search.nixos.org/packages?channel=unstable&from=0&size=50&sort=relevance&query=vscode , then they would find both results. 

True, and this is also the use case for which I care. I'm imagining people clicking on `vscode`, reading the description, installing it, encountering bugs and not knowing that they should have looked at `vscode.fhs`.

jonringer replied:
I was going to add some information to the wiki to improve discoverability anyway. I think we can open a new PR refining description if this does pose some issues for some users.

turion reviewed (changes requested):
Package descriptions can be improved

jonringer commented:
Well, 7 months later, I think this is a good compromise. The original unwrapped workflow hasn't been altered in any fashion, but users now have the option to use extensions freely within vscode.

Extensions which require additional dependencies can be added with the `vscod{e,ium}-fhsWithPackages` to create an fhs environment with additional depedencies (e.g. `rustup`)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/vscod-e-ium-users-you-can-now-use-extensions-without-explicitly-packaging-them-in-nixpkgs/12846/1

CMCDragonkai commented:
If I install `vscode-fhsWithPackages` into my nix-env user profile, does that mean I can install extensions directly from the vscode IDE and not use nix to install extensions? That is, are extensions mutable?

SuperSandro2000 commented:
yes
